persistence/state.go

```diff
@@ -0,0 +1,187 @@
+	flagsMerkleTree
+
+	// Used for iteration purposes only - see https://stackoverflow.com/a/64178235/768439
+	lastMerkleTree
+)
+
+var actorTypeToMerkleTreeName map[types.ActorType]MerkleTree = map[types.ActorType]MerkleTree{
+	types.ActorType_App:  appMerkleTree,
+	types.ActorType_Val:  valMerkleTree,
+	types.ActorType_Fish: fishMerkleTree,
+	types.ActorType_Node: serviceNodeMerkleTree,
+}
+
+var merkleTreeToActorTypeName map[MerkleTree]types.ActorType = map[MerkleTree]types.ActorType{
+	appMerkleTree:         types.ActorType_App,
+	valMerkleTree:         types.ActorType_Val,
+	fishMerkleTree:        types.ActorType_Fish,
+	serviceNodeMerkleTree: types.ActorType_Node,
+}
+
+var actorTypeToSchemaName map[types.ActorType]types.ProtocolActorSchema = map[types.ActorType]types.ProtocolActorSchema{
+	types.ActorType_App:  types.ApplicationActor,
+	types.ActorType_Val:  types.ValidatorActor,
+	types.ActorType_Fish: types.FishermanActor,
+	types.ActorType_Node: types.ServiceNodeActor,
+}
+
+func newMerkleTrees() (map[MerkleTree]*smt.SparseMerkleTree, error) {
+	// We need a separate Merkle tree for each type of actor or storage
+	trees := make(map[MerkleTree]*smt.SparseMerkleTree, int(lastMerkleTree))
+
+	for treeType := MerkleTree(0); treeType < lastMerkleTree; treeType++ {
+		// TODO_IN_THIS_COMMIT: Rather than using `NewSimpleMap`, use a disk based key-value store
+		nodeStore := smt.NewSimpleMap()
+		valueStore := smt.NewSimpleMap()
+
+		trees[treeType] = smt.NewSparseMerkleTree(nodeStore, valueStore, sha256.New())
+	}
+	return trees, nil
+}
+
+func loadMerkleTrees(map[MerkleTree]*smt.SparseMerkleTree, error) {
+	log.Fatalf("loadMerkleTrees not implemented yet")
+}
+
+func (p *PostgresContext) updateStateHash() error {
+	// Update all the merkle trees
+	for treeType := MerkleTree(0); treeType < lastMerkleTree; treeType++ {
+		switch treeType {
+		case appMerkleTree:
+			fallthrough
+		case valMerkleTree:
+			fallthrough
+		case fishMerkleTree:
+			fallthrough
+		case serviceNodeMerkleTree:
+			actorType, ok := merkleTreeToActorTypeName[treeType]
+			if !ok {
+				return fmt.Errorf("no actor type found for merkle tree: %v\n", treeType)
+			}
+			actors, err := p.getActorsUpdatedAtHeight(actorType, p.Height)
+			if err != nil {
+				return err
+			}
+			if err != p.updateActorsTree(actorType, actors) {
+				return err
+			}
+		case accountMerkleTree:
+			fallthrough
+		case poolMerkleTree:
+			fallthrough
+		case blocksMerkleTree:
+			fallthrough
+		case paramsMerkleTree:
+			fallthrough
+		case flagsMerkleTree:
+			// log.Println("TODO: merkle tree not implemented", treeType)
+		default:
+			log.Fatalln("Not handled yet in state commitment update", treeType)
+		}
+	}
+
+	// Get the root of each Merkle Tree
+	roots := make([][]byte, 0)
+	for treeType := MerkleTree(0); treeType < lastMerkleTree; treeType++ {
+		roots = append(roots, p.merkleTrees[treeType].Root())
+	}
+
+	// DISCUSS(drewsky): In #152, we discussed the ordering of the roots
```

Comment:
Consider we could add more trees in the future.

I would encourage to future proof the solution with `strict` ordering using the enum

From what I can tell - it seems like this solution does just that. So I `think` I am a fan of this solution... LMK if I'm missing anything